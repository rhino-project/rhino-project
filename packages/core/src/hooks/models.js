import { isObject } from 'lodash';
import { useMemo } from 'react';
import { getModel } from 'rhino/utils/models';

/**
 * @typedef {import('../utils/models.js').Model} Model
 */

/**
 * Memoize a model by name or with an existing model object
 *
 * @param {string | Model} model - The model name or model to memoize
 * @returns {Model} Memoized model object
 *
 * @example
 *    const model = useModel('blog_post')
 *    const model = useModel(getModel('blog_post'))
 */
export const useModel = (model) =>
  useMemo(() => (isObject(model) ? model : getModel(model)), [model]);
