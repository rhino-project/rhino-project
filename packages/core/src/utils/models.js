import { filter, find, toPath, set } from 'lodash';
import modelLoader from 'rhino/models';

/**
 * @typedef {object} Model
 * @property {string} model
 * @property {string} name
 */

export const getModel = (model) => modelLoader.api.components.schemas[model];
export const getParentModel = (model) => getModel(model.ownedBy);
export const getModelAncestors = (model) => {
  const result = [];
  model = getParentModel(model);
  while (model != null) {
    result.push(model);
    model = getParentModel(model);
  }
  return result;
};
export const getBaseOwnerFilters = (model, baseOwnerId) => {
  const filters = {};
  const pathToRoot = getModelAncestors(model);
  const rootModel = pathToRoot[pathToRoot.length - 1];
  // if the root model is globally owned, we don't want to add any filters. these models have 'global' as parent.
  // only the base owner will have null as parent.
  if (baseOwnerId && rootModel && rootModel.ownedBy == null) {
    const modelNames = pathToRoot.map((el) => el.model);
    set(filters, ['filter', ...modelNames], baseOwnerId);
  }
  return filters;
};

export const getCreatableAttributes = (model) =>
  filter(model.properties, (a) => a.creatable);
export const getUpdatableAttributes = (model) =>
  filter(model.properties, (a) => a.updatable);

export const isReference = (attribute) => attribute.type === 'reference';
export const getReferenceAttributes = (model) =>
  filter(model.properties, (a) => isReference(a));

export const isIdentifier = (attribute) => attribute.type === 'identifier';
export const getIdentifierAttribute = (model) =>
  find(model.properties, (a) => isIdentifier(a));

export const getModuleInfo = (module) =>
  modelLoader.api.info['x-rhino']?.modules?.[module];

export const authOwnerProperty = () => getModuleInfo('rhino')?.authOwner;
export const authOwnerModel = () => getModel(authOwnerProperty());

export const baseOwnerProperty = () => getModuleInfo('rhino')?.baseOwner;
export const baseOwnerModel = () => getModel(baseOwnerProperty());
export const isBaseOwned = (model) => model.ownedBy === baseOwnerProperty();
export const getBaseOwnedModels = () =>
  filter(modelLoader.api.components.schemas, isBaseOwned);

export const getOwnedModels = (model) =>
  filter(modelLoader.api.components.schemas, (m) => m.ownedBy === model.model);

export const getModelFromRef = (attribute) => {
  const ref =
    // Direct ref
    attribute['$ref'] ||
    // anyOf ref, probably because its nullable
    attribute?.anyOf?.[0]?.['$ref'] ||
    // Any array of refs
    attribute?.items?.['$ref'] ||
    // Any array of anyOf refs
    attribute?.items?.anyOf?.[0]?.['$ref'];

  if (!ref) return null;

  const parts = ref.split('/');
  const refName = parts[parts.length - 1];

  const model = modelLoader.api.components.schemas[refName];

  console.assert(
    model,
    `Reference for attribute ${attribute.name} was not found`,
    attribute
  );

  return model;
};

export const getModelAndAttributeFromPath = (model, path) => {
  const split = toPath(path);
  const length = split.length;
  let last_model = model;
  let index = 0;
  let attribute = null;
  let operator = null;
  let plainPath = [];

  while (model && index < length) {
    const [attributeName, operatorName] = split[index++].split('::');

    // A number means its an array indicator
    if (!isNaN(parseInt(attributeName))) {
      continue;
    }
    plainPath.push(attributeName);

    // FIXME Gross hack
    if (attributeName === 'display_name') {
      return [
        last_model,
        {
          model: model.model,
          name: model.model,
          // The last attribute in the path should be the parent model
          readableName: attribute?.readableName || 'Name',
          type: 'string'
        }
      ];
    } else {
      attribute = model.properties[attributeName];
      operator = operatorName;
    }

    last_model = model;
    model = getModelFromRef(attribute);
  }

  return index && index === length
    ? [last_model, attribute, operator, plainPath.join('.')]
    : [undefined, undefined];
};

export const getAttributeFromPath = (model, path) =>
  getModelAndAttributeFromPath(model, path)[1];

export const oauthProviders = () => getModuleInfo('rhino')?.oauth || [];

export const hasModule = (moduleName) =>
  getModuleInfo(moduleName) !== undefined;

export const hasOrganizationsModule = () => hasModule('rhino_organizations');
export const hasNotificationsModule = () => hasModule('rhino_notifications');
export const hasSubscriptionsModule = () => hasModule('rhino_subscriptions');
