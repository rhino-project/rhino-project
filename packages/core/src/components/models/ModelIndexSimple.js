import PropTypes from 'prop-types';

import ModelIndexProvider from './ModelIndexProvider';
import { useModelIndexController } from '../../hooks/controllers';
import { Spinner } from 'reactstrap';
import { useMemo } from 'react';

export const ModelIndexSimple = ({ children, fallback = false, ...props }) => {
  const controller = useModelIndexController({
    ...props
  });
  const { isInitialLoading } = controller;

  // Fallback mirrors React 18 Suspense
  const renderFallback = useMemo(() => {
    if (!isInitialLoading || !fallback) return children;

    if (fallback === true) return <Spinner />;

    return fallback;
  }, [children, fallback, isInitialLoading]);

  return (
    <ModelIndexProvider {...controller}>{renderFallback}</ModelIndexProvider>
  );
};

ModelIndexSimple.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ])
};

export default ModelIndexSimple;
