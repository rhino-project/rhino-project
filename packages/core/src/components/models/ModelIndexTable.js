import {
  cloneElement,
  isValidElement,
  useCallback,
  useEffect,
  useMemo,
  useState
} from 'react';
import PropTypes from 'prop-types';
import {
  createColumnHelper,
  getCoreRowModel,
  useReactTable
} from '@tanstack/react-table';

import {
  useGlobalComponentForModel,
  useOverrides
} from 'rhino/hooks/overrides';

import { useBaseOwnerNavigation } from 'rhino/hooks/history';
import { useModelIndexContext } from 'rhino/hooks/controllers';
import { filter, isString } from 'lodash';
import { usePaths } from 'rhino/hooks/paths';
import Table from '../table/Table';
import ModelCell from './ModelCell';
import ModelHeader from './ModelHeader';
import ModelFooter from './ModelFooter';
import { getModelShowPath } from '../../utils/routes';
import ModelSection from './ModelSection';

const getViewablePaths = (model) =>
  filter(model.properties, (a) => {
    return (
      a.type !== 'identifier' &&
      a.name !== model.ownedBy &&
      a.type !== 'array' &&
      a.type !== 'jsonb' &&
      a.type !== 'text' &&
      // a.type !== 'reference' &&
      !a.name.endsWith('_attachment')
    );
  }).map((a) => a.name);

const defaultComponents = {
  ModelHeader,
  ModelCell,
  ModelFooter
};

const isDesc = (order) => order?.charAt(0) === '-';

const getSortableAttributes = (model) =>
  filter(
    model.properties,
    (a) =>
      a.type === 'string' ||
      a.type === 'datetime' ||
      a.type === 'float' ||
      a.type === 'integer'
  );

export const ModelIndexTableBase = ({ overrides, ...props }) => {
  const { ModelHeader, ModelCell, ModelFooter } = useOverrides(
    defaultComponents,
    overrides
  );

  const { model, order, resources, results, setOrder } = useModelIndexContext();
  const { baseRoute, paths, sortPaths } = props;
  const baseOwnerNavigation = useBaseOwnerNavigation();
  const [sorting, setSorting] = useState([]);

  const pathsOrDefault = useMemo(() => {
    if (props.overrides?.ModelTable?.props?.paths)
      console.warn('ModelTable pass legacy paths prop');

    return (
      paths ||
      // Legacy
      props.overrides?.ModelTable?.props?.paths ||
      getViewablePaths(model)
    );
  }, [paths, props.overrides?.ModelTable?.props?.paths, model]);

  const computedPaths = usePaths(pathsOrDefault, resources);

  const handleRowClick = useCallback(
    (row) =>
      baseOwnerNavigation.push(
        `${baseRoute}${getModelShowPath(model, row.original.id)}`
      ),
    [baseRoute, baseOwnerNavigation, model]
  );

  const sortable = useMemo(
    () => sortPaths || getSortableAttributes(model).map((a) => a.name),
    [sortPaths, model]
  );

  const columns = useMemo(
    () =>
      computedPaths.map((path, idx) => {
        if (isValidElement(path)) {
          const accessor =
            path.props?.accessor ||
            (isString(path.props?.path) ? path.props?.path : null);
          // FIXME: Any issue using idx as id?
          const id = path.props?.id || idx.toString();
          const header =
            path.props?.header ||
            (() => (
              <ModelHeader model={model} path={path?.props?.path || null} />
            ));
          const cell = (props) => cloneElement(path, { model, ...props });
          const footer =
            path.props?.footer ||
            (() => (
              <ModelFooter model={model} path={path?.props?.path || null} />
            ));

          if (accessor) {
            return columnHelper.accessor(accessor, {
              id,
              header,
              cell,
              footer,
              enableSorting: false
            });
          }

          return columnHelper.display({
            id,
            header,
            cell,
            footer,
            enableSorting: false
          });
        }

        // Path is a string
        const cell = (info) => (
          <ModelCell model={model} path={path} {...info} />
        );

        const header = (info) => (
          <ModelHeader model={model} path={path} {...info} />
        );
        const footer = (info) => (
          <ModelFooter model={model} path={path} {...info} />
        );

        return columnHelper.accessor(path, {
          id: path,
          header,
          cell,
          footer,
          enableSorting: sortable.includes(path),
          enableMultiSort: sortable.includes(path)
        });
      }),
    [computedPaths, model, sortable]
  );

  useEffect(() => {
    // The initial pass will set the sorting state from the order prop
    if (sorting.length === 0) {
      setSorting(
        order.split(',').map((order) => {
          const id = order.replace('-', '');
          return { id, desc: isDesc(order) };
        })
      );

      return;
    }

    // The sorting state has changed, update the order in the controller
    setOrder(
      sorting.map((order) => (order.desc ? '-' + order.id : order.id)).join(',')
    );
  }, [order, setOrder, sorting]);

  const table = useReactTable({
    data: results || [],
    columns,
    getCoreRowModel: getCoreRowModel(),
    enableMultiSort: true,
    enableSortingRemoval: false,
    manualSorting: true,
    state: {
      sorting
    },
    onSortingChange: setSorting
  });

  return (
    <ModelSection baseClassName="index-table">
      <Table table={table} onRowClick={handleRowClick} {...props} />
    </ModelSection>
  );
};

ModelIndexTableBase.propTypes = {
  baseRoute: PropTypes.string.isRequired,
  overrides: PropTypes.object
};

ModelIndexTableBase.defaultProps = {
  baseRoute: ''
};

const columnHelper = createColumnHelper();

export const ModelIndexTable = (props) =>
  useGlobalComponentForModel('ModelIndexTable', ModelIndexTableBase, props);

export default ModelIndexTable;
