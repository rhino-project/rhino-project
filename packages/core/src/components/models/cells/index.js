export * from './ModelCellArray';
export * from './ModelCellArrayReference';
export * from './ModelCellAttachment';
export * from './ModelCellAttachmentDownload';
export * from './ModelCellAttachmentImage';
export * from './ModelCellAttachments';
export * from './ModelCellBoolean';
export * from './ModelCellBooleanIcon';
export * from './ModelCellCountry';
export * from './ModelCellCurrency';
export * from './ModelCellDate';
export * from './ModelCellDateTime';
export * from './ModelCellDateTimeDistance';
export * from './ModelCellEnum';
export * from './ModelCellFloat';
export * from './ModelCellInteger';
export * from './ModelCellLinkEmail';
export * from './ModelCellLinkTelephone';
export * from './ModelCellReference';
export * from './ModelCellString';
export * from './ModelCellTime';
export * from './ModelEditableCellBoolean';
export * from './ModelEditableCellReference';
export * from './ModellCellIdentifier';
