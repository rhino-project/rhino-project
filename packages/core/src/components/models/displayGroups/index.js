export * from './ModelDisplayGroupArray';
export * from './ModelDisplayGroupArrayReference';
export * from './ModelDisplayGroupAttachment';
export * from './ModelDisplayGroupAttachmentImage';
export * from './ModelDisplayGroupAttachments';
export * from './ModelDisplayGroupBoolean';
export * from './ModelDisplayGroupCurrency';
export * from './ModelDisplayGroupDate';
export * from './ModelDisplayGroupDateTime';
export * from './ModelDisplayGroupEnum';
export * from './ModelDisplayGroupFloat';
export * from './ModelDisplayGroupInteger';
export * from './ModelDisplayGroupReference';
export * from './ModelDisplayGroupString';
export * from './ModelDisplayGroupText';
export * from './ModelDisplayGroupTime';
