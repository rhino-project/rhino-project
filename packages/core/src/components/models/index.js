export * from './ModelCell';
export * from './ModelCreate';
export * from './ModelCreateActions';
export * from './ModelCreateForm';
export * from './ModelCreateHeader';
export * from './ModelCreateModal';
export * from './ModelCreateModalActions';
export * from './ModelCreateProvider';
export * from './ModelCreateSimple';
export * from './ModelDisplayGroup';
export * from './ModelDisplayGroupFloating';
export * from './ModelDisplayGroupHorizontal';
export * from './ModelDisplayLabel';
export * from './ModelEdit';
export * from './ModelEditActions';
export * from './ModelEditForm';
export * from './ModelEditHeader';
export * from './ModelEditModal';
export * from './ModelEditModalActions';
export * from './ModelEditProvider';
export * from './ModelEditSimple';
export * from './ModelFieldGroup';
export * from './ModelFieldGroupFloating';
export * from './ModelFieldGroupHorizontal';
export * from './ModelFieldLabel';
export * from './ModelFilter';
export * from './ModelFilterGroup';
export * from './ModelFilterLabel';
export * from './ModelFilters';
export * from './ModelFiltersProvider';
export * from './ModelFiltersSimple';
export * from './ModelFooter';
export * from './ModelHeader';
export * from './ModelIndex';
export * from './ModelIndexActions';
export * from './ModelIndexCardGrid';
export * from './ModelIndexHeader';
export * from './ModelIndexProvider';
export * from './ModelIndexSimple';
export * from './ModelIndexTable';
export * from './ModelPager';
export * from './ModelProvider';
export * from './ModelSearch';
export * from './ModelSection';
export * from './ModelShow';
export * from './ModelShowActions';
export * from './ModelShowDescription';
export * from './ModelShowHeader';
export * from './ModelShowProvider';
export * from './ModelShowRelated';
export * from './ModelShowSimple';
