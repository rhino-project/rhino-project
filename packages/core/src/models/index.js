import { each } from 'lodash';

import staticModels from 'models/static';

const MODEL_PATH = '/api/info/openapi';

const hoistRhino = (data) => {
  each(Object.keys(data.components.schemas), (schema) => {
    const model_uplifted = {
      ...data.components.schemas[schema]['x-rhino-model']
    };

    data.components.schemas[schema] = {
      ...data.components.schemas[schema],
      ...model_uplifted
    };

    // OpenAPI does required on the model level, we need it on the
    // attribute level
    const required = data.components.schemas[schema].required;

    each(
      Object.keys(data.components.schemas[schema].properties),
      (property) => {
        const attribute_uplifted = {
          ...data.components.schemas[schema].properties[property][
            'x-rhino-attribute'
          ]
        };

        if (required?.includes(property)) {
          attribute_uplifted['x-rhino-required'] = true;
        }

        data.components.schemas[schema].properties[property] = {
          ...data.components.schemas[schema].properties[property],
          ...attribute_uplifted
        };
      }
    );
  });

  return data;
};

class ModelLoader {
  async loadModels(forceStatic = false) {
    if (forceStatic || import.meta.env.PROD) {
      this.api = hoistRhino(staticModels);
    } else {
      const { networkApiCall } = await import('rhino/lib/networking');

      return networkApiCall(MODEL_PATH).then(
        (response) => (this.api = hoistRhino(response.data))
      );
    }
  }
}

export default new ModelLoader();
