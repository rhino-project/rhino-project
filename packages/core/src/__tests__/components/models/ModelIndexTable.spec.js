import { render } from '@testing-library/react';
import { sharedModelTests } from './sharedModelTests';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { MemoryRouter } from 'react-router-dom';
import { ModelIndexTable } from '../../../components/models/ModelIndexTable';
import { ModelIndexSimple } from '../../../components/models/ModelIndexSimple';

describe('ModelIndexTable', () => {
  const Foo = () => <div>Foo</div>;
  const Bar = () => <div>Bar</div>;
  const Baz = () => <div>Baz</div>;

  const wrapper = ({ children }) => {
    const queryClient = new QueryClient({
      defaultOptions: {
        queries: {
          retry: false
        }
      }
    });

    return (
      <QueryClientProvider client={queryClient}>
        <MemoryRouter>
          <ModelIndexSimple
            model="user"
            fallback={false}
            queryOptions={{ enabled: false }}
          >
            {children}
          </ModelIndexSimple>
        </MemoryRouter>
      </QueryClientProvider>
    );
  };

  sharedModelTests(ModelIndexTable);

  it(`should allow local overrides`, async () => {
    const { asFragment } = render(
      <ModelIndexTable
        overrides={{
          ModelHeader: Foo,
          ModelCell: Bar,
          ModelFooter: Baz
        }}
      />,
      { wrapper }
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it(`should allow local overrides of Table`, async () => {
    const { asFragment } = render(
      <ModelIndexTable
        overrides={{
          Table: Foo
        }}
      />,
      { wrapper }
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
