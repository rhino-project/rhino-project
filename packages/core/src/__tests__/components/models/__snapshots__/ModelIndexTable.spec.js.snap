// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ModelIndexTable > should allow local overrides 1`] = `
<DocumentFragment>
  <div
    class="model-index-table model-index-table-user"
  >
    <table
      class="table table-striped table-hover"
    >
      <thead>
        <tr>
          <th
            class="table-header-clickable"
          >
            <div
              class="d-flex flex-row gap-1"
            >
              <div>
                Foo
              </div>
              <div
                class="d-flex flex-column ms-auto"
              >
                <svg
                  fill="currentColor"
                  height="12"
                  width="12"
                >
                  <use
                    xlink:href="/@fs/Users/jpr/src/Rhino/opensource-construction/rhino-project/node_modules/.pnpm/bootstrap-icons@1.11.3/node_modules/bootstrap-icons/bootstrap-icons.svg#caret-up"
                  />
                </svg>
                <svg
                  fill="currentColor"
                  height="12"
                  width="12"
                >
                  <use
                    xlink:href="/@fs/Users/jpr/src/Rhino/opensource-construction/rhino-project/node_modules/.pnpm/bootstrap-icons@1.11.3/node_modules/bootstrap-icons/bootstrap-icons.svg#caret-down"
                  />
                </svg>
              </div>
            </div>
          </th>
          <th
            class=""
          >
            <div
              class="d-flex flex-row gap-1"
            >
              <div>
                Foo
              </div>
            </div>
          </th>
          <th
            class=""
          >
            <div
              class="d-flex flex-row gap-1"
            >
              <div>
                Foo
              </div>
            </div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="table-row-clickable"
        >
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
        </tr>
        <tr
          class="table-row-clickable"
        >
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
        </tr>
        <tr
          class="table-row-clickable"
        >
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
        </tr>
        <tr
          class="table-row-clickable"
        >
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
        </tr>
        <tr
          class="table-row-clickable"
        >
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
        </tr>
        <tr
          class="table-row-clickable"
        >
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
        </tr>
        <tr
          class="table-row-clickable"
        >
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
        </tr>
        <tr
          class="table-row-clickable"
        >
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
        </tr>
        <tr
          class="table-row-clickable"
        >
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
        </tr>
        <tr
          class="table-row-clickable"
        >
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
          <td>
            <div>
              Bar
            </div>
          </td>
        </tr>
      </tbody>
      <tfoot>
        <tr>
          <th>
            <div>
              Baz
            </div>
          </th>
          <th>
            <div>
              Baz
            </div>
          </th>
          <th>
            <div>
              Baz
            </div>
          </th>
        </tr>
      </tfoot>
    </table>
  </div>
</DocumentFragment>
`;

exports[`ModelIndexTable > should allow local overrides of Table 1`] = `
<DocumentFragment>
  <div
    class="model-index-table model-index-table-user"
  >
    <div>
      Foo
    </div>
  </div>
</DocumentFragment>
`;

exports[`ModelIndexTable > should render with global override shorthand 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;

exports[`ModelIndexTable > should render with global override shorthand for model 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;
